content/posts/2018-10-21--complex-command-handler-in-javascript.md

```diff
@@ -0,0 +1,352 @@
+
+Let's start with `handleInParallel` and `Promise`. Our method needs to return an instance of `Promise` that should be resolved once all provided commands are completed. To execute these commands in parallel we can call `handle` method of `SimpleCommandHandler` for each of them. This way we'll obtain a set of promises. When all of them are resolved we can resolve the promise that was our return value. Luckily we can use `Promise.all` method to easily obtain the composite promise we need to return.
+
+```js
+class CompositeCommandHandler {
+	// Methods defined previously.
+
+    handleInParallel(commands) {
+        const promises = commands.map(command => {
+            return this.handle(command);
+        });
+
+        return Promise.all(promises);
+    }
+
+    handleInSequence(commands) {
+        // To be defined.
+    }
+}
+```
+
+First we map each element of `commands` to a `Promise` using `handle` method recursively (as command can be either simple or composite and we've already implemented the logic of handling them there). Then this array of promisses is used as an argument for `Promise.all` which produces composite instance of `Promise` class that we need to return.
+
+We need only one more method to make our `CompositeCommandHandler` work and it's `handleInSequence` Let's try to use `Promise` again.
+
+What we need to do is to handle the first element of `commands` argument. Only when it is completed (e.g. promise resulting from handling it is resolved) we can handle  the next element of `commands`. We should do so until all commands are handled and then we can use the last instance of `Promise` as our return value. Let's express this with code then --- for the sake of simplicity we can assume that `commands` will never be an empty array.
+
+```js
+class CompositeCommandHandler {
+    // Methods defined previously.
+	
+    handleInSequence(commands) {
+        let lastPromise = this.handle(commands[0]);
+        for (const command of commands.slice(1)) {
+            lastPromise = promise.then(() => this.handle(command));
+        }
+        
+        return lastPromise;
+    }
+}
+```
+
+In this example we don't care about the result of handling each command and therefore we can simple chain promises using `then` callbacks that don't require any arguments. We produce first promise outside of the `for ... of` loop and then for each subsequent command we add a `then` callback that will handle the net element of `commands` and return resulting promise. This creates a new promise that we store in the `lastPromise` variable and use it in next loop. The final value of `lastPromise` is used as return value as it will be resolved when the last command is completed.
```

Comment:
```suggestion
In this example, we don't care about the result of handling each command and therefore we can simply chain promises using `then` callbacks that don't require any arguments. We produce the first promise outside of the `for ... of` loop and then for each subsequent command we add a `then` callback that will handle the net element of `commands` and return resulting promise. This creates a new promise that we store in the `lastPromise` variable and use it in the next loop. The final value of `lastPromise` is used as return value as it will be resolved when the last command is completed.
```